### PR TITLE
ENG-18700: Introduced transient readers that don't participate in PBD…

### PR DESCRIPTION
**Pull request by manjujames**
… data deletion.

Additional bookkeeping for transient readers when assertions are on.

### CODE REVIEW

File: tests/frontend/org/voltdb/utils/TestPBDMultipleReaders.java [hunk 1]
```diff
@@ -240,6 +240,52 @@ public void testSegmentClosingWriterMultipleReaders() throws Exception {
         assertEquals(1, m_pbd.numberOfSegments());
     }
 
+    @Test
+    public void testTransientReader() throws Exception {
```
Comment (rdykiel) [inline tests/frontend/org/voltdb/utils/TestPBDMultipleReaders.java]: Please comment on the logic of your test case and the expected results at each significant test step.
Also add a test case showing that a transient reader doesn't delete the PBD buffers.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 2]
```diff
@@ -96,10 +96,16 @@ public synchronized void discard() {
         // If a rewind occurred this is set to the segment id where this cursor was before the rewind
         private long m_rewoundFromId = -1;
         private boolean m_cursorClosed = false;
+        private boolean m_isTransient;
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Sure

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 3]
```diff
@@ -1306,6 +1315,20 @@ private void closeAndDeleteSegment(PBDSegment<M> segment) throws IOException {
             m_usageSpecificLog.debug("Closing and deleting segment " + segment.file()
                 + " (final: " + segment.isFinal() + ")");
         }
+        if (assertionsOn) {
+            for (ReadCursor reader : m_readCursors.values()) {
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: OK.

File: tests/frontend/org/voltdb/utils/TestPBDMultipleReaders.java (hunk 1 above)
Comment (manjujames) [reply tests/frontend/org/voltdb/utils/TestPBDMultipleReaders.java]: OK

File: src/frontend/org/voltdb/utils/BinaryDeque.java [hunk 4]
```diff
@@ -95,13 +95,25 @@
 
     /**
      * Start a BinaryDequeReader for reading, positioned at the start of the deque.
+     * This is equivalent to {@link BinaryDeque#openForRead(cursorId, isTransient)} with false for <code>isTransient</code>.
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/BinaryDeque.java]: OK, thanks.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 5]
```diff
@@ -523,6 +529,9 @@ private boolean canDeleteSegment(PBDSegment<M> segment) throws IOException {
             }
 
             for (ReadCursor cursor : m_readCursors.values()) {
+                if (cursor.m_isTransient) {
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Not sure about that. Do we need to keep track of open transient readers? Should we allow opening a transient reader with the same name again? If we don't track them, there can be multiple open with the same name. 

May be we can revisit this once we start using it with topic streams.

**Pull request merged**